bluetooth: host: LE CS subevent result reassembly

bugo-nordic opened the pull request:
Adds HCI support for LE CS subevent result continue event and the reassembly logic for the partial results.

sean-madigan commented on subsys/bluetooth/host/cs.c:
```diff
@@ -360,73 +543,181 @@ void bt_hci_le_cs_subevent_result(struct net_buf *buf)
+		/* We can now clear the any reassembly buffer allocated for this procedure,
+		 * to avoid code duplication, we're using the abort label to do so
+		 */
+		goto abort;
```
if we abort he do we double unref? line 624 and 638?

bugo-nordic replied:
Good catch, need to set it back to NULL

sean-madigan replied:
or move this branch above line 623?

bugo-nordic replied:
I think as a practice, it's better to set it to NULL when freed.

sean-madigan commented on subsys/bluetooth/host/cs.c:
```diff
@@ -561,6 +856,13 @@ void bt_hci_le_cs_test_end_complete(struct net_buf *buf)
+	struct net_buf *reassembly_buf = get_reassembly_buf(BT_HCI_LE_CS_TEST_CONN_HANDLE, false);
+
+	if (reassembly_buf) {
+		LOG_WRN("De-allocating a dangling reassembly buffer");
```
Will it also be left dangling on a disconnect? Maybe need to add a disconnected callback to handle this?

sean-madigan commented on subsys/bluetooth/host/cs.c:
```diff
@@ -22,6 +22,189 @@ LOG_MODULE_REGISTER(bt_cs);
+	}
+
+	if (reassembly_buf->len) {
+		LOG_WRN("Over-written incomplete CS subevent results");
```
In cases like this should we raise the callback to the app for the previous result? Or is this not expected to happen

bugo-nordic replied:
Normally, we allow partial results when the cs procedure or subevent is aborted. But in that case, the controller reports this by setting the subevent and procedure done status bits accordingly.

What we detect here is an abnormal use case where the controller sends some partial results followed by a new subevent results without marking the previous ones as done/aborted. 

I'm not sure reporting this data is necessary because there is no way to verify its integrity.

sean-madigan commented on subsys/bluetooth/host/cs.c:
```diff
@@ -360,72 +543,184 @@ void bt_hci_le_cs_subevent_result(struct net_buf *buf)
+			goto abort;
+		}
+
+		free_reassembly_buf(&reassembly_buf);
```
Should we not free reassembly buf on line 696 after we reset it? Reassembly buff is only valid for a single subevent rather than a procedure, correct?

sean-madigan replied:
I guess both probably work

bugo-nordic replied:
It's allocated per procedure, otherwise we'll be allocating and deallocating redundantly for the common use case. Do you think allocating per subevent is better?

sean-madigan replied:
I think it's fine, I was just confused

sean-madigan commented on subsys/bluetooth/host/cs.c:
```diff
@@ -561,6 +856,13 @@ void bt_hci_le_cs_test_end_complete(struct net_buf *buf)
+	struct net_buf *reassembly_buf = get_reassembly_buf(BT_HCI_LE_CS_TEST_CONN_HANDLE, false);
+
+	if (reassembly_buf) {
+		LOG_WRN("De-allocating a dangling reassembly buffer");
```
Controller should send pending CS subevent results to host in this case before we get the CS test end event

bugo-nordic replied:
Discussed offline: Will leave this check here as it's detecting an invalid behavior. Will add something similar to CS Procedure Enable Complete event when it's added.

olivier-le-sage commented on subsys/bluetooth/host/cs.c:
```diff
@@ -360,72 +543,184 @@ void bt_hci_le_cs_subevent_result(struct net_buf *buf)
+		}
+
+		p_result = &reassembled_result;
+		p_result->step_data_buf = &reassembly_buf->b;
```
is it ok to access b directly ? I saw it's labeled INTERNAL_HIDDEN

olivier-le-sage commented on subsys/bluetooth/host/cs.c:
```diff
@@ -360,72 +543,184 @@ void bt_hci_le_cs_subevent_result(struct net_buf *buf)
+	}
 
-		notify_cs_subevent_result(conn, &result);
+	uint16_t step_data_len = evt->num_steps_reported ? buf->len : 0;
```
Not sure if num_steps_reported == 0 makes sense in a continue event. The spec seems to contradict itself a bit

`The Num_Steps_Reported parameter indicates the number of CS steps for which
results are reported. The Step_Mode[i] parameter indicates the CS mode for each CS
step, and the Step_Channel[i] parameter indicates the channel used for each CS step.
A Controller may return a value 0 when a subevent is aborted.`

but then

`For each CS procedure in a repeat sequence, the Controller shall report the results
for each CS step in each CS subevent of the CS procedure even if the CS step has been aborted for any reason.`

bugo-nordic replied:
The wording for subevent result event is:

> The HCI_LE_CS_Subevent_Result event shall be generated when the local Controller has results to report for a CS subevent during the CS procedure.

To me, this suggests that a controller doesn't have to wait until a subevent is finished to start reporting the results. Such a controller can send some partial results (whilst the procedure is running) but then send a continuation event to report that subevent/procedure was aborted with 0 remaining steps.

bugo-nordic commented on subsys/bluetooth/host/cs.c:
```diff
@@ -360,72 +543,184 @@ void bt_hci_le_cs_subevent_result(struct net_buf *buf)
+	reassembled_result.header.subevent_done_status = evt->subevent_done_status;
+	reassembled_result.header.procedure_abort_reason = evt->procedure_abort_reason;
+	reassembled_result.header.subevent_abort_reason = evt->subevent_abort_reason;
+	reassembled_result.header.num_antenna_paths = evt->num_antenna_paths;
```
It's not 100% clear what should the host do with this information as it won't be part of the ranging data. Following the assumption that, we could receive events where this can be zero or non-zero, we'll accept whatever non-zero value passed.

Though looks like I forgot to remove this line. Will do that now.

sean-madigan commented on subsys/bluetooth/host/cs.c:
```diff
@@ -22,6 +22,230 @@ LOG_MODULE_REGISTER(bt_cs);
+		return NULL;
+	}
+
+	if (!disconn_cb_registered) {
```
Might be simpler just to do 
```
BT_CONN_CB_DEFINE(cs_conn_callbacks) = {
	.disconnected = clear_on_disconnect,
};
```
Instead somewhere in the file?

bugo-nordic replied:
Of course!

olivier-le-sage commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -640,6 +640,10 @@ struct bt_conn_le_cs_subevent_result {
 		/** Number of CS steps in the subevent.
 		 */
 		uint8_t num_steps_reported;
+		/** Step number, on which the subevent was aborted
```
we should also add some docs here https://github.com/zephyrproject-rtos/zephyr/pull/79497/files#diff-4538e90f4e61dd34802aa1c0a5382dc89d4df6032ec6a3dbbf4eef746ebdd488R622 about how abort_step relates to the status

For instance I expect that if 200 steps are reported like this:
- subevent result: 100 steps, subevent_done_status = 0x1
- subevent result continue: 99 steps, subevent_done_status = 0x1
- subevent result continue: 1 step, subevent_done_status = 0xF

this struct will have subevent_done_status = 0xF and step_data_buf pointing to 200 step data elements.
So we should clarify in the API that the user should interpret this as "199 non-aborted steps + 1 failed/aborted step".

olivier-le-sage commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -640,6 +640,10 @@ struct bt_conn_le_cs_subevent_result {
 		 */
 		uint8_t num_steps_reported;
+		/** Step number, on which the subevent was aborted
+		 *  255 if subevent was not aborted
```
```suggestion
		 *  255 if subevent_done_status is @ref BT_CONN_LE_CS_SUBEVENT_COMPLETE
```

we can consider removing BT_CONN_LE_CS_SUBEVENT_INCOMPLETE since this is no longer exposed to the user

jhedberg commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -3636,6 +3636,20 @@ struct bt_hci_evt_le_cs_subevent_result {
+	uint8_t config_id;
+	uint8_t procedure_done_status;
+	uint8_t subevent_done_status;
+	uint8_t procedure_abort_reason: 4;
+	uint8_t subevent_abort_reason: 4;
```
This seems risky. I don't think it's safe to make assumptions of bitfield in-memory ordering, i.e. it's something that will depend on the architecture that this gets built for. I'd therefore suggest to use just a single `uint8_t` and provide helper macros to access the two sub-octet fields.

bugo-nordic replied:
It looks like there are multiple examples of bitfield access to protocol data in zephyr:

https://github.com/zephyrproject-rtos/zephyr/blob/52540d8412630edb2c64f4fcca1c1165bcce1ac3/subsys/bluetooth/controller/ll_sw/pdu.h#L417-L438

We have quite a few bitfields for the CS step data so this approach would be better for the user instead of access macros. So I suggest, we modify our bitfields similarly.

jhedberg replied:
@bugo-nordic @olivier-le-sage I just had to comment on the same thing here: https://github.com/zephyrproject-rtos/zephyr/pull/75637#discussion_r1800017520

Based on the text from the C standard, even though in practice the endianness-based variants may work, this is not something that's universally always guaranteed to be safe.

olivier-le-sage replied:
I agree with you that the C standard doesn't define this, but the C standard also doesn't define that a byte is 8 bits or a pointer represents an address, or has size X on an X-bit CPU. It intentionally leaves these up to the ABI.

For instance if you search for bitfield or bit-field in the following documents:
* ARM EABI section 7.1.7 Bit-fields  (https://web.eecs.umich.edu/%7Eprabal/teaching/resources/eecs373/ARM-AAPCS-EABI-v2.08.pdf)
* i386 ABI section 3-6 (https://www.sco.com/developers/devspecs/abi386-4.pdf)
* x86-64 (https://raw.githubusercontent.com/wiki/hjl-tools/x86-psABI/x86-64-psABI-1.0.pdf)
* MIPS section 3.7 (https://refspecs.linuxfoundation.org/elf/mipsabi.pdf)
* RISCV section 2.1 (https://d3s.mff.cuni.cz/files/teaching/nswi200/202324/doc/riscv-abi.pdf)

I think you'll agree that any compiler for these targets would produce correct assembly from the Zephyr link layer code linked above.

Here is another example from the linux kernel: https://github.com/torvalds/linux/blob/master/include/uapi/linux/ip.h

jhedberg replied:
@olivier-le-sage fair enough, and thanks for the very good set of ABI references (which must have taken some time to collect!). I still think it's healthy to acknowldege the assumptions we're making, i.e. the correctness of the code is based on specific ABI definitions rather than the C standard. Personally, I tend to go for helper macros in this kind of situations, but I agree that from a code readability perspective it's nice to have bitfields in the structs. Anyway, since we have precedents in Zephyr as well as the Linux kernel, I'm ok with continuing with the endianness-based struct variant approach.

@gzh-terry based on the above, do you want to revise the approach you're taking in #75637?

bugo-nordic committed:
bluetooth: host: LE CS subevent result reassembly

Adds HCI support for LE CS subevent result continue
event and the reassembly logic for the partial results.

When subevent results are completed or the subevent is
aborted, the user callback is invoked with a buffer
pointing to the HCI event buffer, so no copy is done.

When subevent results are incomplete, then a reassembly
buffer is allocated from a fixed sized pool. This buffer
is used for the reassembling of the subevent result
containing all of the step data, which is then passed
to the user via the callback. kconfigs have been added
to set the size and the count of the reassembly buffer.

Signed-off-by: Burak Gorduk <burak.gorduk@nordicsemi.no>

bugo-nordic committed:
bluetooth: host Accommodate platform endianness for CS types

Modifies the CS bitfields that are used for accessing HCI
event data to take platform endiannes into consideration.

Signed-off-by: Burak Gorduk <burak.gorduk@nordicsemi.no>
